src/controllers/addressBook/addressBook.ts

```diff
@@ -0,0 +1,160 @@
+import { getAddress } from 'ethers'
+
+import { Storage } from '../../interfaces/storage'
+import EventEmitter from '../eventEmitter/eventEmitter'
+
+export type Contacts = Array<{
+  name: string
+  address: string
+  isWalletAccount?: boolean
+}>
+
+/**
+ * Address Book controller
+ */
+
+export class AddressBookController extends EventEmitter {
+  // Manually added contact (stored in storage)
+  #manuallyAddedContacts: Contacts = []
+
+  // Contacts, generated on the fly from the accounts in the wallet (not stored in storage)
+  #accountsInWalletContacts: Contacts = []
+
+  #storage: Storage
+
+  constructor(storage: Storage) {
+    super()
+
+    this.#storage = storage
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    this.#load()
+  }
+
+  set accountsInWalletContacts(accountsInWalletContacts: Contacts) {
+    this.#accountsInWalletContacts = accountsInWalletContacts.map((contact) => ({
+      ...contact,
+      isWalletAccount: true
+    }))
+    this.emitUpdate()
+  }
```

Comment:
This mechanism for setting the accounts in wallet contacts (via the ADDRESS_BOOK_CONTROLLER_SET_ACCOUNTS_IN_WALLET_CONTACTS action from the extension) can be improved.

Since the MainController creates the AddressBookController, the MainController can pass the `MainController.accounts` in here. The accounts array in there keeps the ref, so that it will always be in sync if the accounts in the extension change.

Let's discuss this further offline.